src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -0,0 +1,1490 @@
+    /// To avoid this we also create ephemeral child before starting recursive removal.
+    /// (The existence of child node does not allow to remove parent node).
+    Coordination::Requests ops;
+    Coordination::Responses responses;
+    fs::path my_keeper_path = keeper_path;
+    String drop_lock_path = my_keeper_path / "dropped" / "lock";
+    ops.emplace_back(zkutil::makeRemoveRequest(my_keeper_path / "replicas", -1));
+    ops.emplace_back(zkutil::makeCreateRequest(my_keeper_path / "dropped", "", zkutil::CreateMode::Persistent));
+    ops.emplace_back(zkutil::makeCreateRequest(drop_lock_path, "", zkutil::CreateMode::Ephemeral));
+    Coordination::Error code = my_keeper->tryMulti(ops, responses);
+
+    if (code == Coordination::Error::ZNONODE || code == Coordination::Error::ZNODEEXISTS)
+    {
+        LOG_WARNING(log, "Table {} is already started to be removing by another replica right now", replica_path);
+    }
+    else if (code == Coordination::Error::ZNOTEMPTY)
+    {
+        LOG_WARNING(log, "Another replica was suddenly created, will keep the table {}", replica_path);
+    }
+    else if (code != Coordination::Error::ZOK)
+    {
+        zkutil::KeeperMultiException::check(code, ops, responses);
+    }
+    else
+    {
+        auto drop_lock = zkutil::EphemeralNodeHolder::existing(drop_lock_path, *my_keeper);
+        LOG_INFO(log, "Removing table {} (this might take several minutes)", keeper_path);
+        removeTableNodesFromZooKeeper(my_keeper, drop_lock);
+    }
+}
+
+std::optional<StorageKafka2::TopicPartitionLocks>
+StorageKafka2::lockTopicPartitions(zkutil::ZooKeeper & keeper_to_use, const TopicPartitions & topic_partitions)
+{
+    std::vector<fs::path> topic_partition_paths;
+    topic_partition_paths.reserve(topic_partitions.size());
+    for (const auto & topic_partition : topic_partitions)
+        topic_partition_paths.emplace_back(getTopicPartitionPath(topic_partition));
+
+    Coordination::Requests ops;
+
+    static constexpr auto ignore_if_exists = true;
+
+    for (const auto & topic_partition_path : topic_partition_paths)
+    {
+        const auto lock_file_path = String(topic_partition_path / lock_file_name);
+        LOG_TRACE(log, "Creating locking ops for: {}", lock_file_path);
+        ops.push_back(zkutil::makeCreateRequest(topic_partition_path, "", zkutil::CreateMode::Persistent, ignore_if_exists));
+        ops.push_back(zkutil::makeCreateRequest(lock_file_path, kafka_settings->kafka_replica_name.value, zkutil::CreateMode::Ephemeral));
+    }
+    Coordination::Responses responses;
+
+    if (const auto code = keeper_to_use.tryMulti(ops, responses); code != Coordination::Error::ZOK)
+    {
+        if (code != Coordination::Error::ZNODEEXISTS)
+            zkutil::KeeperMultiException::check(code, ops, responses);
+
+        // Possible optimization: check the content of logfiles, if we locked them, then we can clean them up and retry to lock them.
```

Comment:
So this (ZNODEEXISTS) can happen if multiple consumers got the same asignment? Why can this happen? (I am not aware kafka internals)

Reply:
This can happen during rebalance: e.g.: `consumerA` already got the new assignment, tries to lock the partition, but another `consumerB` is still busy inserting an already pulled batch. In this case `consumerA` will got `ZNODEEXISTS` error, because `consumerB` still holds the lock.